webapi/src/main/scala/org/knora/webapi/responders/v2/resources/CheckObjectClassConstraints.scala

```diff
@@ -36,72 +39,82 @@ object CheckObjectClassConstraints {
+      objectConstraints <- propertysObjectConstraint(propertyIri, entityInfo, propertyInfo, ontologyRepo)
 
-      // Get the property's object class constraint. If this is a link value property, we want the object
-      // class constraint of the corresponding link property instead.
-      propertyInfoForObjectClassConstraint = if (propertyInfo.isLinkValueProp)
-                                               entityInfo.propertyInfoMap(propertyIri.fromLinkValuePropToLinkProp)
-                                             else
-                                               propertyInfo
+      // Check each value.
+      _ <-
+        ZIO.foreachDiscard(valuesToCreate.map(_.valueContent)) {
+          case linkValueContentV2: LinkValueContentV2 =>
+            // It's a link value.
+            for {
+              _ <-
+                ZIO.when(!propertyInfo.isLinkValueProp) {
+                  propertyRequiresValue(resourceIdForErrorMsg, propertyIri, objectConstraints)
+                }
+
+              // Does the resource that's the target of the link belongs to a subclass of the
+              // link property's object class constraint?
+              linkTargetClass     = linkTargetClasses(linkValueContentV2.referredResourceIri)
+              linkTargetClassInfo = entityInfo.classInfoMap(linkTargetClass)
+              _ <-
+                ZIO.when(!linkTargetClassInfo.allBaseClasses.exists(objectConstraints.contains)) {
+                  // No. If the target resource already exists, use its IRI in the error message.
+                  // Otherwise, use the client's ID for the resource.
+                  ZIO.fail(
+                    propertyInvalid(
+                      resourceIdForErrorMsg,
+                      linkValueContentV2,
+                      clientResourceIDs,
+                      propertyIri,
+                      objectConstraints,
+                    ),
+                  )
+                }
+            } yield ()
+
+          case other: ValueContentV2 =>
+            // It's not a link value. Check that its type is equal to the property's object class constraint.
+            ZIO.when(!objectConstraints.contains(other.valueType)) {
+              propertyRequiresValue(resourceIdForErrorMsg, propertyIri, objectConstraints)
+            }
+        }
+    } yield ()
 
-      propertyIriForObjectClassConstraint = propertyInfoForObjectClassConstraint.entityInfoContent.propertyIri
+  private def propertysObjectConstraint(
+    propertyIri: SmartIri,
+    entityInfo: EntityInfoGetResponseV2,
+    propertyInfo: ReadPropertyInfoV2,
+    ontologyRepo: OntologyRepo,
+  )(implicit
+    stringFormatter: StringFormatter,
+  ): Task[List[SmartIri]] = {
+    // * Represents an OWL property definition as returned in an API response.
+    // Get the property's object class constraint. If this is a link value property, we want the object
+    // class constraint of the corresponding link property instead.
```

Comment:
**suggestion (non-blocking):** this comment might be more helpful if it were proper scaladoc, which would make it clear *what* "represents an OWL property definition" etc.